### DISCUSSION TITLE
.ipynb notebooks?

### DISCUSSION THREAD

**Discussion by bgraham89** [TheAlgorithms/Python#8695] [Ideas]
**What's your thoughts on adding .pynb files into the mix?**

One difference between using .ipynb compared to .py files for algorithms, is the visual flexibility provided by markdown cells in .ipynb files. Here's an example of a markdown cell accompanying an algorithm. 

![markdown_render](https://user-images.githubusercontent.com/81972602/234808476-66ff1ab4-5f27-471c-9275-622dac0eeebb.png)

Benefits of markdown cells include:
* Flexible layouting with support for headers
* The markdown cells support mathjax, so the inherent maths related to an algorithm can be written simply, and displayed nicely. 

![markdown_raw](https://user-images.githubusercontent.com/81972602/234808486-36df4c12-d976-4de0-be7b-44d44c46cbc0.png)

There are also other benefits of .ipynb files, related to the possibility of including separate python cells:

* It's possible to have python cells that output visualisations
* It's possible to have python cells dedicated to parts of an algorithm.

I think all of these benefits contribute to educational scaffolding, which is in the spirit of this repository. I don't suggest replacing the .py files, as scripts themselves are executable and easy to use, but a section dedicated to .ipynb files is a way to provide more than just a script, in case anybody wants to learn more. It's worth pointing out that .ipynb files can include hyperlinks too which can be lead to further resources.

Just an idea, what are your thoughts on it?

**Comment by Youssef-AMARZOU** [comment]
This is exactly what I mean - seeing Horner's algorithm with MathJax rendering the formula alongside the code makes it way more educational than just a .py file. Students can see the math and the implementation side by side.

One suggestion though: keep the .py files as the main implementation for testing and code quality, and use the notebooks as supplementary material. That way we get the best of both worlds.

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Jupyter

**Reply by bgraham89** [reply]
That doesn't look like it's algorithm centred, but yeah, it shows us what a jupyter notebook collection is.
